src/editor.cls.php

```diff
@@ -0,0 +1,155 @@
+      'side',                                   // Context
+      'high'                                    // Priority
+    );
+  }
+
+  /**
+  * Display the post meta box.
+  *
+  * @since    4.3
+  */
+  function meta_box_html( $post ) {
+    if ( false ) {
+      $post_meta = get_post_meta( $post->ID, 'litespeed_metabox_vpi', true );
+      $desktop_organized_post_meta = '';
+      $mobile_organized_post_meta = '';
+      $desktop_warning_msg = 'Desktop VPI is empty';
+      $mobile_warning_msg = 'Mobile VPI is empty';
+      if ( isset( $post_meta['desktop'] ) && is_array( $post_meta['desktop'] ) ) {
+        $desktop_organized_post_meta = esc_attr( implode("\n", $post_meta['desktop'] ) );
+        $desktop_warning_msg = '';
+      }
+      if ( isset( $post_meta['mobile'] ) && is_array( $post_meta['mobile'] ) ) {
+        $mobile_organized_post_meta = esc_attr( implode("\n", $post_meta['mobile'] ) );
+        $mobile_warning_msg = '';
+      }
+      wp_nonce_field( basename( __FILE__ ), 'litespeed_metabox_nonce' );
+
+      require_once LSCWP_DIR . 'tpl/editor/sidebar-metabox.tpl.php';
+    }
+  }
+
+  /**
+  * Save the meta box’s metadata.
+  *
+  * @since    4.3
+  */
+  function save_meta( $post_id, $post ) {
+    /* Verify the nonce before proceeding. */
+    if ( !isset( $_POST['litespeed_metabox_nonce'] ) || !wp_verify_nonce( $_POST['litespeed_metabox_nonce'], basename( __FILE__ ) ) )
+      return $post_id;
+
+    /* Get the post type object. */
+    $post_type = get_post_type_object( $post->post_type );
+
+    /* Check if the current user has permission to edit the post. */
+    if ( !current_user_can( $post_type->cap->edit_post, $post_id ) )
+      return $post_id;
+
+    if ( false ) {
+      Debug2::debug( '[Editor] save_meta' );
+
+      /* Get the meta key. */
+      $meta_key = 'litespeed_metabox_vpi';
+
+      /* Get the meta value of the custom field key. */
+      $meta_value = get_post_meta( $post_id, $meta_key, true );
+
+      /* Get the posted data and sanitize it for use as an HTML class. */
+      $new_desktop_meta_value = isset( $_POST['vpi_desktop'] ) ? $_POST['vpi_desktop'] : '';
+      $new_mobile_meta_value = isset( $_POST['vpi_mobile'] ) ? $_POST['vpi_mobile'] : '';
+
+      // if there is no VPI record and the user has not given any thing and updated,
+      //meta_value is '', and both new_desktop_meta_value and new_desktop_meta_value is ''
+      if ( ( $meta_value == $new_desktop_meta_value ) && ( $meta_value == $new_desktop_meta_value ) ) {
+        // don't bother
+        Debug2::debug( '[Editor] no row added and changed for vpi' );
+        return;
+      }
+
+      /* Clean up the meta data from posted by the user, check if empty, if so then make it into an empty array. */
+      /* Desktop */
+      $new_desktop_meta_value = trim( $new_desktop_meta_value, ' ' );
+      $new_desktop_meta_value = ( $new_desktop_meta_value == '' ) ? '' : explode( "\n", $new_desktop_meta_value );
```

Comment:
redundant parentheses

Reply:
You don't need `trim` before `explode`. You can always explode first. This way, it is an array, hence you don't need line 108 to convert it to array.
As the begin/end/middle lines may contain extra empty lines, you can use `array_filter` to drop them. Also always use `array_unique` to drop duplicated lines.